### DISCUSSION TITLE
Vehicle capacity of nodes and edges

### DISCUSSION THREAD

**Discussion by frothm** [openTCS/opentcs#32] [Q&A] [answered]
### Background:
I am using openTCS with a simple TCP communication adapter to control virtual robots in a discrete-event simulation. The transportation network layout in openTCS was realized using "halt positions" for intersections as well as load handling stations and simple paths between these positions. All paths are directed; bidirectional sections have two paths and two beginning/end points each.

### Issue:
In my simulation, each edge of the layout has a capacity of like 5 to 10 vehicles, i.e. several vehicles can move on the same edge between two intersection points. OpenTCS on the other hand expects each "halt position" to have a capacity of 1 by default. Edges don't have any capacity. Therefore the openTCS Kernel only sends a MovementCommand to a vehicle if the next position in its route is empty.

This behavior of openTCS yields a vast reduction of the whole system's transport capacity. Also it easily produces deadlocks, as vehicles often cannot finish their orders due to waiting in front of an occupied node.

### Question:
Can I overcome this problem by somehow raising vehicle capacities? Or is there any other helpful feature in openTCS?

Some thoughts on possible workarounds, but not sure about effort and effectiveness:
* Use "reporting positions" instead of "halt positions" in the openTCS layout.
* Split each edge into a chain of halt/reporting positions according to its capacity.
* Give PrecisePosition updates additionally to the Position updates so that the vehicles quickly leave the intersection positions.

I would be glad if anyone had a hint about a useful feature that I missed, or an idea for a smart workaround :)

**Comment by swltr** [comment, marked as answer]
You're right: points (nodes in your graph) have a capacity of 1, and openTCS's scheduler implementation does not allow multiple vehicles to travel the same path (edges in your graph) at the same time. One reason for this is that having multiple vehicles on the same path complicates things in edge cases like initialization. (If multiple vehicles reported to be on path _p_ when the system is started, openTCS would need to have a mechanism to find out in which order they are queued up there. This would need to be considered when assigning new orders to them - assigning an order to a vehicle somewhere in the middle would be a bad choice.)

What will not help is marking points as reporting positions and/or providing precise position updates. These are basically only helpers for visualization between halt points and are not taken into account by openTCS's traffic management.

Your second idea is the default solution / workaround / way of handling this: have shorter paths, which unfortunately means creating more points and paths in the plant model. Alternatively, it would be possible to create a modified implementation of openTCS's `Scheduler` interface which handles capacities > 1 and use that instead of the default one that openTCS comes with.

A note about deadlocks, merely because it's unclear to me what exact situation you are referring to and to avoid misunderstandings: a deadlock requires circular waiting. If vehicle _v1_ waits at point _p1_ for vehicle _v2_ to leave point _p2_, but _v2_ doesn't leave because it doesn't have an order to go somewhere else, that's not a deadlock. If _v2_, on the other hand, is waiting for _v1_ to leave _p1_, that's a classic deadlock. The latter can happen, of course, especially in models where there can be circles with a small number of points. One way to handle this is - again - to keep it in mind when creating the driving course model. Another way would be to integrate an algorithmic solution. Contributions that solve this in a generic way _and_ are actually usable in real-world plant scenarios are certainly welcome! :-)
